subsys/bluetooth/host/hci_core.c

```diff
@@ -9330,7 +9330,6 @@ int bt_pub_key_gen(struct bt_pub_key_cb *new_cb)
 		return -ENOTSUP;
 	}
 
-	new_cb->_next = pub_key_cb;
```

Comment:
This removes the linked list property of the callbacks, and makes it impossible to have multiple pending callbacks for the pubkey generation. As far as I know, we'd like to be able to call pubkey gen from multiple modules and still get the callbacks in each of them once it's ready.

If we really don't want to have multiple callbacks, we can just pass the function pointer here instead, and get rid of the for loop in `le_pkey_complete`, but I think we need it. I think a better solution would be to scan through the list or use the `_next` pointer to determine whether it's already in the list. Perhaps use sys/slist.h in this case.

Reply:
@joerchan pointed out this issue: #23292.
It's currently not possible to freeze or otherwise ensure sole control of the ECC keypair in Zephyr, mostly because of limitations in HCI. Joakim's second proposal matches the changes in this PR, where only a single module can have its callback listed at a time. If we want to go in this direction, we should probably return `-EBUSY` if there's already a callback present, so some other module doesn't come in and steal the callback away from us.